### PR TITLE
Downstream changes to avoid race condition in ZkStateReader on watched collection states

### PR DESCRIPTION
**Pull request by patsonluk**
## Description
The fix and jira issues on apache/solr:
https://github.com/apache/solr/pull/909 and https://github.com/apache/solr/pull/966
https://issues.apache.org/jira/browse/SOLR-16257

We are going to downstream this change to our release/8.8 branch

CI test: https://app.circleci.com/pipelines/github/cowpaths/fs-solr/1685/workflows/230b9f7d-aff8-4e6b-8096-b48437a9a8cc

### CODE REVIEW

File: solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java
Comment (hiteshk25) [inline solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java:746]: is it some specific condition here?

File: solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java
Comment (hiteshk25) [inline solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java:357]: computeIfPresent, will compute if key presents? then below first two `if's` not required?
Also, is computeIfPresent thread safe?

File: solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java
Comment (patsonluk) [reply solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java:746]: Yes. This is to maintain the consistency with the old behavior, which a null state wuold have no key in `watchedCollectionStates` https://github.com/fullstorydev/lucene-solr/pull/176/files/44b45f0d3ef418815299961f8269b283bdf9a207#diff-c797d60b04b72ceaf84395b8a77c9c51299259d4214e58bcd92331c2bb9c51acL2079 . Hence not inserted into the result

File: solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java
Comment (patsonluk) [reply solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java:357]: I assume u refer to the `oldState== null` check? 

`watch` is non null, but `watch.currentState` (which is `oldState`) can be null (that means someone is watching a collection, but such collection not yet exists in the zk), hence the checks are required. 

Yes, `ConcurrentHashMap#computeIfPresent` is thread safe as quoted
>The entire method invocation is performed atomically. 

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#computeIfPresent-K-java.util.function.BiFunction-

**Review by hiteshk25 — APPROVED** [review]
LGTM (Thanks @patsonluk please do some testing playpen before merging - cluster restart, node restart, overseer node restart, solr unit tests etc)

**Comment by patsonluk** [comment]
> LGTM (Thanks @patsonluk please do some testing playpen before merging - cluster restart, node restart, overseer node restart, solr unit tests etc)

Testing on playpen done:

Deployed the pending fs-solr PR https://github.com/cowpaths/fs-solr/commit/a5d84161bfd1b331fb8af74a76a2698993f5b23b to playpen
image


After first restart, solrman started reporting some weird problem on c08-1,3,5. for collection prs11024 (replica status). Upon inspection in the zk. It's found that each replica has 2 PRS entries, such as:
```
...
core_node991:0:A:L
core_node991:58:A:L
core_node993:0:A:L
core_node993:49:A:L
core_node995:0:A:L
core_node995:40:A:L
core_node998:0:A:L
core_node998:44:A:L
core_node99:0:A:L
core_node99:43:A:L
core_node9:0:A:L
core_node9:43:A:L
```

Not sure how it arrives at that stage, after restarting those nodes, the problems are gone. But the duplicated entries remain.

in order to confirm that this PR does not cause such state, Created a brand new PRS collection with 1k shard (patson-big-prs) and restart the nodes again, and there's no problem for such new collection

2 data node ran into restart problem from solrsidekick,
```
{"level":"INFO","time":"2022-08-25T18:42:08.391Z","caller":"fs/skeleton/auth.go:57","message":"fs/skeleton/auth.go:279 [httpMachineAuth] failed to machine auth PUT call to /stab?blocking=true&threaddump=false\n...caused by fs/auth/machine/auth.go:171 [validateJWT] invalid JWT\n...caused by fs/gcloud/iam/claims.go:31 [skewTolerantClaims.Valid] token used before issued by 34s, allowed skew is 30s","org_id":"","request_path":"/stab","request_id":"solr-c08-4:527590:1661422791615441332"}
```
 but it's likely not related to the solr change. restarting solrsidekick solved the issue


Tested 5 splits and 5 moves on PRS collection `o-HNS-na1` . Verified the operations completed as expected

**Comment by patsonluk** [comment]
Holding off merge on this until after the Solr deployment on Sep 1!

**Comment by hiteshk25** [comment]
> > LGTM (Thanks @patsonluk please do some testing playpen before merging - cluster restart, node restart, overseer node restart, solr unit tests etc)
> 
> Testing on playpen done:
> 
> Deployed the pending fs-solr PR [cowpaths/fs-solr@a5d8416](https://github.com/cowpaths/fs-solr/commit/a5d84161bfd1b331fb8af74a76a2698993f5b23b) to playpen image
> 
> After first restart, solrman started reporting some weird problem on c08-1,3,5. for collection prs11024 (replica status). Upon inspection in the zk. It's found that each replica has 2 PRS entries, such as:
> 
> ```
> ...
> core_node991:0:A:L
> core_node991:58:A:L
> core_node993:0:A:L
> core_node993:49:A:L
> core_node995:0:A:L
> core_node995:40:A:L
> core_node998:0:A:L
> core_node998:44:A:L
> core_node99:0:A:L
> core_node99:43:A:L
> core_node9:0:A:L
> core_node9:43:A:L
> ```
> 
> Not sure how it arrives at that stage, after restarting those nodes, the problems are gone. But the duplicated entries remain.
> 
> in order to confirm that this PR does not cause such state, Created a brand new PRS collection with 1k shard (patson-big-prs) and restart the nodes again, and there's no problem for such new collection
> 
> 2 data node ran into restart problem from solrsidekick,
> 
> ```
> {"level":"INFO","time":"2022-08-25T18:42:08.391Z","caller":"fs/skeleton/auth.go:57","message":"fs/skeleton/auth.go:279 [httpMachineAuth] failed to machine auth PUT call to /stab?blocking=true&threaddump=false\n...caused by fs/auth/machine/auth.go:171 [validateJWT] invalid JWT\n...caused by fs/gcloud/iam/claims.go:31 [skewTolerantClaims.Valid] token used before issued by 34s, allowed skew is 30s","org_id":"","request_path":"/stab","request_id":"solr-c08-4:527590:1661422791615441332"}
> ```
> 
> but it's likely not related to the solr change. restarting solrsidekick solved the issue
> 
> Tested 5 splits and 5 moves on PRS collection `o-HNS-na1` . Verified the operations completed as expected

We never seen duplicate prs entries in zk. Wondering collection creation bug fix(2nd merge) caused this?

**Comment by patsonluk** [comment]
For reference, as discussed with @hiteshk25 , the collection that has duplicated PRS entries, `prs11024` was probably created long time ago and might be hard to trace if any tested operations might have created such weird state. 

Since we cannot reproduce the same issue with newly created PRS collection, we will keep watching but assume this is not caused by the recent changes

**Pull request merged**
